From f413ce43e5cf248c4b2c9ea17a0e145e577760c6 Mon Sep 17 00:00:00 2001
From: Rob Latham <robl@mcs.anl.gov>
Date: Fri, 8 Nov 2013 16:25:54 -0600
Subject: [PATCH] make MPIU_Assert more robust

instead of assigning and then checking for overflow, check if the
overflow will happen, and then assign.
---
 src/include/mpiutil.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/include/mpiutil.h b/src/include/mpiutil.h
index f2fbaf04..d6e764de 100644
--- a/src/include/mpiutil.h
+++ b/src/include/mpiutil.h
@@ -239,9 +239,9 @@ int MPIR_Assert_fail_fmt(const char *cond, const char *file_name, int line_num,
 #endif
 #define MPIU_Assign_trunc(dst_,src_,dst_type_)                                         \
     do {                                                                               \
-        dst_ = (dst_type_)(src_);                                                      \
         /* will catch some of the cases if the expr_inttype macros aren't available */ \
-        MPIU_Assert((dst_) == (src_));                                                 \
+        MPIU_Assert((src_) == (dst_type_)(src_));                                      \
+        dst_ = (dst_type_)(src_);                                                      \
     } while (0)
 
 /*
-- 
1.8.3.2

